src/data/sequence.rs

```diff
@@ -0,0 +1,727 @@
+        )
+    }
+
+    /// Returns the data shell - that is - everything except the Values themselves.
+    pub fn shell(&self, expected_data_version: impl Into<Version>) -> Result<Self> {
+        let expected_data_version = to_absolute_version(
+            expected_data_version.into(),
+            self.expected_data_version() as usize,
+        )
+        .ok_or(Error::NoSuchEntry)? as u64;
+
+        let access_list = self
+            .access_list
+            .iter()
+            .filter(|a| a.expected_data_version() <= expected_data_version)
+            .cloned()
+            .collect();
+
+        let owners = self
+            .owners
+            .iter()
+            .filter(|owner| owner.expected_data_version <= expected_data_version)
+            .cloned()
+            .collect();
+
+        Ok(Self {
+            address: self.address,
+            data: Vec::new(),
+            access_list,
+            owners,
+            _flavour: self._flavour,
+        })
+    }
+
+    /// Return a value for the given Version (if it is present).
+    pub fn get(&self, version: Version) -> Option<&Value> {
+        let absolute_version = to_absolute_version(version, self.data.len())?;
+        self.data.get(absolute_version)
+    }
+
+    /// Return the current data entry (if it is present).
+    pub fn current_data_entry(&self) -> Option<DataEntry> {
+        match self.data.last() {
+            Some(value) => Some(DataEntry::new(self.data.len() as u64, value.to_vec())),
+            None => None,
+        }
+    }
+
+    /// Get a range of values within the given versions.
+    pub fn in_range(&self, start: Version, end: Version) -> Option<Values> {
+        let range = to_absolute_range(start, end, self.data.len())?;
+        Some(self.data[range].to_vec())
+    }
+
+    /// Return all Values.
+    pub fn values(&self) -> &Values {
+        &self.data
+    }
+
+    /// Get owner at version.
+    pub fn owner_at(&self, version: impl Into<Version>) -> Option<&Owner> {
+        let version = to_absolute_version(version.into(), self.owners.len())?;
+        self.owners.get(version)
+    }
+
+    /// Returns history of all owners
+    pub fn owner_history(&self) -> Vec<Owner> {
+        self.owners.clone()
+    }
+
+    /// Get history of owners within the range of versions specified.
+    pub fn owner_history_range(&self, start: Version, end: Version) -> Option<Vec<Owner>> {
+        let range = to_absolute_range(start, end, self.owners.len())?;
+        Some(self.owners[range].iter().copied().collect())
+    }
+
+    /// Get access control at version.
+    pub fn access_list_at(&self, version: impl Into<Version>) -> Option<&C> {
+        let version = to_absolute_version(version.into(), self.access_list.len())?;
+        self.access_list.get(version)
+    }
+
+    /// Returns history of all access list states
+    pub fn access_list_history(&self) -> Vec<C> {
+        self.access_list.clone()
```

Comment:
Could we avoid this clone here?
If we need it then we can add it later but it's preferred that we pass references.

Reply:
Thing is, these places, when returning reference in original implementation, they clone at next level of usage (in `safe_vault`).
So I thought we might just as well do that from start.

Reply:
OK, I had no complaints from clippy in safe_vault about it.
It got more complicated at that level though (it seemed). When I did this I mostly went for the first seemingly OK-ish thing that compiled. Didn't want to spend too much time on details that are easily sorted out by a rust coder.

In case you check out the safe_vault PR you can see the usage there, and if you know how it should be dealt with there.
Otherwise I'll have another look at it in a bit.

Reply:
It's okay to `.clone()` in safe_vault and other layers. As I mentioned before, by restricting it at the lower layer, we avoid the always-clone that would happen in this case. i.e. we wouldn't be aware of it. I've seen places where two getters are available for both the clone and reference. That way the user of the API is aware of the extra memory usage. I'll take a look at the usage in safe_vault, thanks!